src/DDpackage.cpp

```diff
@@ -3,1355 +3,775 @@
+                e.w = cn.mulCached(x.w, y.w);
 
-            sum.r += cv.r;
-            sum.i += cv.i;
+                if (CN::equalsZero(e.w)) {
+                    cn.releaseCached(e.w);
+                    return mEdge::zero;
+                }
+                return e;
+            }
         }
 
-        r   = DDzero;
-        r.w = cn.getTempCachedComplex(sum.r, sum.i);
+        std::array<mEdge, NEDGE> edge{};
+        mEdge                    e1{}, e2{};
+        for (auto i = 0U; i < NEDGE; i += RADIX) {
+            for (auto j = 0U; j < RADIX; j++) {
+                edge[i + j] = mEdge::zero;
+                for (auto k = 0U; k < RADIX; k++) {
+                    if (!x.isTerminal() && x.p->v == var) {
+                        e1 = x.p->e[i + k];
+                    } else {
+                        e1 = xCopy;
+                    }
+                    if (!y.isTerminal() && y.p->v == var) {
+                        e2 = y.p->e[j + RADIX * k];
+                    } else {
+                        e2 = yCopy;
+                    }
+
+                    auto m = multiply2(e1, e2, static_cast<Qubit>(var - 1));
+
+                    if (k == 0 || edge[i + j].w == CN::ZERO) {
+                        edge[i + j] = m;
+                    } else if (m.w != CN::ZERO) {
+                        auto old_e  = edge[i + j];
+                        edge[i + j] = add2(edge[i + j], m);
+                        cn.releaseCached(old_e.w);
+                        cn.releaseCached(m.w);
+                    }
+                }
+            }
+        }
+        e = makeDDNode(var, edge, true);
 
-        CTinsert(xCopy, yCopy, r, fid);
-        CN::mul(r.w, r.w, x.w);
-        CN::mul(r.w, r.w, y.w);
+        matrixMultiplication.insert(xCopy, yCopy, {e.p, e.w});
 
-        return {CN::val(r.w.r), CN::val(r.w.i)};
+        if (e.w != CN::ZERO && (x.w != CN::ONE || y.w != CN::ONE)) {
+            if (e.w == CN::ONE) {
+                e.w = cn.mulCached(x.w, y.w);
+            } else {
+                CN::mul(e.w, e.w, x.w);
+                CN::mul(e.w, e.w, y.w);
+            }
+            if (CN::equalsZero(e.w)) {
+                cn.releaseCached(e.w);
+                return mEdge::zero;
+            }
+        }
+        return e;
     }
 
-    ComplexValue Package::innerProduct(const Edge& x, const Edge& y) {
+    ComplexValue Package::innerProduct(const Package::vEdge& x, const Package::vEdge& y) {
         if (x.p == nullptr || y.p == nullptr || CN::equalsZero(x.w) || CN::equalsZero(y.w)) { // the 0 case
             return {0, 0};
         }
 
         [[maybe_unused]] const auto before = cn.cacheCount;
-        short                       w      = x.p->v;
+
+        auto w = x.p->v;
         if (y.p->v > w) {
             w = y.p->v;
         }
-        const ComplexValue ip = innerProduct(x, y, w + 1);
+        const auto ip = innerProduct(x, y, static_cast<Qubit>(w + 1));
```

Comment:
I'd kept the explicit type. `innerProduct` could reasonably return a `Complex` type.

Reply:
Agreed 👍🏼